### PR TITLE
Se adjunta la práctica3_2.py. Es necesario instalar scikit-learn y sk…

### PR DESCRIPTION
**Pull request by GilbertoBarranco**
…learn

### CODE REVIEW

File: GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    return [word2features(sent, i) for i in range(len(sent))]
+
+
+def sent2labels(sent):
+    return [label for token, label in sent]
+
+
+# Comenzamos a tratar el corpus.
+
+import ast
```
Comment (umoqnier) [inline GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py:112]: Este import no se utiliza O:

File: GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+from sklearn.metrics import f1_score
+
+print("\tF1-score:", f1_score(y_pred_flat, y_test_flat, average="macro"))
+
+from sklearn.metrics import ConfusionMatrixDisplay
+
+disp = ConfusionMatrixDisplay.from_predictions(y_test_flat, y_pred_flat)
+disp.figure_.suptitle("Confusion Matrix")
+print(f"Confusion matrix:\n{disp.confusion_matrix}")
```
Comment (umoqnier) [inline GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py:181]: <3

File: GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+# Prepare data for CRF
+X = [[word2features(sent, i) for i in range(len(sent))] for sent in corpus]
+y = [[map_tag(pos) for _, pos in sent] for sent in corpus]
+
+# Split the data into training and testing sets
+X_train, X_test, y_train, y_test = train_test_split(
+    X, y, test_size=0.2, random_state=42
+)
+
+from inspect import Attribute
```
Comment (umoqnier) [inline GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py:130]: Esto no se utiliza

File: GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+corpus = open("corpus_otomi", "r")
+corpus = corpus.read()
+corpus = re.sub(r"\n", ", ", corpus)
+corpus = eval(corpus)
+corpus = get_tagged_sentences(corpus)
+
+# Prepare data for CRF
+X = [[word2features(sent, i) for i in range(len(sent))] for sent in corpus]
+y = [[map_tag(pos) for _, pos in sent] for sent in corpus]
```
Comment (umoqnier) [inline GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py:123]: ¿Por qué decidiste mapear las tags a su forma expandida?

File: GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+
+def map_tag(tag):
+    try:
+        return postags[tag]
+    except:
+        return "unkwn"
+
+
+def word2features(sent, i):
```
Comment (umoqnier) [inline GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py:59]: Curioso, obtienes un accuracy de casi 100%. Lo que me hace pensar que algo anda raro o mal. Creo que esta cayendo en overfit el modelo, pero no encuentro el problema. Curioso en verdad

File: GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    word = sent[i][0]
+    postag = sent[i][1]
+
+    features = {
+        "word.lower()": word.lower(),
+        "word[-3:]": word[-3:],
+        "word[-2:]": word[-2:],
+        "word.istitle()": word.istitle(),
+        "postag": postag,
+        "postag[:2]": postag[:2],
```
Comment (umoqnier) [inline GilbertoBarrancoSanchez/practica3_2/práctica_3_2.py:69]: Creo que puede ir por acá, que estás incluyendo las etiquetas POS en las _feature functions_. Creo que con texto nuevo no tendrá buen desempeño en absoluto

**Review by umoqnier — APPROVED** [review]
En principio parece que funciona. Sin embargo, al incluir las etiquetas POS en las feature functions se cae en [overfiting](https://en.wikipedia.org/wiki/Overfitting)  

Al incluir las etiquetas POS en tus *feature functions* rompe el paradigma de aprendizaje supervisado. Pues está entrenando pasándole explícitamente la información de la etiqueta que quiere predecir por lo que el modelo no está generalizando realmente.

Calificación: 6
